internal/exec/stages/files/filesystemEntries.go

```diff
@@ -205,99 +195,43 @@ func depth(n types.Node) uint {
 // mapEntriesToFilesystems builds a map of filesystems to files. If multiple
 // definitions of the same filesystem are present, only the final definition is
 // used. The directories are sorted to ensure /foo gets created before /foo/bar.
-func (s stage) mapEntriesToFilesystems(config types.Config) (map[types.Filesystem][]filesystemEntry, error) {
-	filesystems := map[string]types.Filesystem{}
-	for _, fs := range config.Storage.Filesystems {
-		filesystems[fs.Name] = fs
-	}
-
-	entryMap := map[types.Filesystem][]filesystemEntry{}
-
-	// Sort directories to ensure /a gets created before /a/b.
-	sortedDirs := config.Storage.Directories
-	sort.Stable(ByDirectorySegments(sortedDirs))
+func (s stage) getOrderedCreationList(config types.Config) []filesystemEntry {
+	entries := []filesystemEntry{}
 
 	// Add directories first to ensure they are created before files.
-	for _, d := range sortedDirs {
-		if fs, ok := filesystems[d.Filesystem]; ok {
-			entryMap[fs] = append(entryMap[fs], dirEntry(d))
-		} else {
-			s.Logger.Crit("the filesystem (%q), was not defined", d.Filesystem)
-			return nil, ErrFilesystemUndefined
-		}
+	for _, d := range config.Storage.Directories {
+		entries = append(entries, dirEntry(d))
 	}
 
 	for _, f := range config.Storage.Files {
-		if fs, ok := filesystems[f.Filesystem]; ok {
-			entryMap[fs] = append(entryMap[fs], fileEntry(f))
-		} else {
-			s.Logger.Crit("the filesystem (%q), was not defined", f.Filesystem)
-			return nil, ErrFilesystemUndefined
-		}
+		entries = append(entries, fileEntry(f))
```

Comment:
We intentionally are creating specified directories before files, should we also attempt to create the parent directories of all files & links before we run through file/link creation?

Reply:
we sort them from shallowest to deepest, so directories of files should get created before the files themselves. one TODO is resolving the symlinks at the start of the stage and ordering based on the resolved symlink path rather than the specified path, but I was planning on that in a follow up PR.